mixer/adapter/svcctrl/svcctrl.go

```diff
@@ -12,115 +12,170 @@
+		}
 	}
 
-	op.MetricValueSets = []*sc.MetricValueSet{
-		{
-			MetricName:   "serviceruntime.googleapis.com/api/producer/request_count",
-			MetricValues: []*sc.MetricValue{&metricValue},
-		},
+	if b.config.ServiceConfigs == nil || len(b.config.ServiceConfigs) == 0 {
+		result = multierror.Append(result, errors.New("ServiceConfigs is nil or empty"))
+	} else {
+		for _, cfg := range b.config.ServiceConfigs {
+			if cfg.MeshServiceName == "" || cfg.GoogleServiceName == "" {
+				result = multierror.Append(result,
+					errors.New("MeshServiceName and GoogleServiceName must be non-empty"))
+			}
+
+			if cfg.QuotaConfigs != nil {
+				for _, qCfg := range cfg.QuotaConfigs {
+					if qCfg.QuotaName == "" {
+						result = multierror.Append(result, errors.New("QuotaName is empty"))
+					}
+					if qCfg.Expiration == nil {
+						result = multierror.Append(result, errors.New("quota expiration is nil"))
+					} else {
+						expiration, err := pbtypes.DurationFromProto(qCfg.Expiration)
+						if err != nil {
+							result = multierror.Append(result, err)
+						} else if expiration <= 0 {
+							result = multierror.Append(
+								result, fmt.Errorf(
+									`quota must have postive expiration, but get %v`, expiration))
+						}
+					}
+				}
+			}
+		}
 	}
 
-	reportReq := &sc.ReportRequest{
-		Operations: []*sc.Operation{op},
+	if result != nil {
+		return &adapter.ConfigErrors{Multi: result.(*multierror.Error)}
 	}
-	return reportReq, nil
+	return nil
 }
 
-func (h *handler) Close() error {
-	h.serviceControlClient = nil
-	return nil
+// Build builds an adapter handler.
+func (b *builder) Build(context context.Context, env adapter.Env) (adapter.Handler, error) {
+	client, err := newClient(b.config.CredentialPath)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, err := initializeHandlerContext(env, b.config, client)
+	if err != nil {
+		return nil, err
+	}
+	ctx.checkDataShape = b.checkDataShape
+	ctx.reportDataShape = b.reportDataShape
+	return newHandler(ctx)
 }
 
-////////////////// Config //////////////////////////
+func initializeHandlerContext(env adapter.Env, adapterCfg *config.Params,
+	client serviceControlClient) (*handlerContext, error) {
+	consumerProjCache, err := newLRUCache(
+		int(adapterCfg.RuntimeConfig.CheckCacheSize),
+		toDuration(adapterCfg.RuntimeConfig.CheckResultExpiration))
+	if err != nil {
+		return nil, err
+	}
+
+	configIndex := make(map[string]*config.ServiceConfig)
+	for _, cfg := range adapterCfg.ServiceConfigs {
+		configIndex[cfg.MeshServiceName] = cfg
+	}
+
+	return &handlerContext{
+		env:                env,
+		config:             adapterCfg,
+		serviceConfigIndex: configIndex,
+		consumerProjCache:  consumerProjCache,
+		client:             client,
+	}, nil
+}
 
 // GetInfo registers Adapter with Mixer.
 func GetInfo() adapter.Info {
 	return adapter.Info{
 		Name:        "svcctrl",
-		Impl:        "istio.io/istio/mixer/adapter/svcctrl",
+		Impl:        "istio.io/istio/mixer/adapter/serviceControl",
```

Comment:
this should be the name of the directory. Since the dir name is svcctrl, consider renaming this to the correct dir name